File: experiments/st-johns-weather-map/api/weather_api/layer_identity.py
"""Explicit layer declarations, without title guessing or new provider reads."""
from datetime import datetime
from typing import Iterable

from .models import ImageryAvailability, LayerFieldMapping, catalogue_key_for


def mappings(pairs: Iterable[tuple[str, str | None]]) -> dict:
    """The caller supplies authoritative source/declared-field associations.

    Only the existing field catalogue alias rule resolves a field. A known
    source with an unmapped field remains partial, never an invented bundle key.
    """
    records = []
    seen = set()
    for source, native in pairs:
        key = catalogue_key_for(native) if isinstance(native, str) else None
        identity = (source, key, native)
        if identity in seen:
            continue
        seen.add(identity)
        records.append(LayerFieldMapping(source_id=source, field_key=key, declared_field=native))
    status = 'known' if records and all(row.field_key is not None for row in records) else 'partial' if records else 'unknown'
    return dict(field_mappings=records, mapping_status=status,
                mapping_reason='Explicit source/field declaration; this is not proof of a retrieved point value' if status == 'known'
                else 'Source is explicit; one or more declared fields have no catalogue association' if records
                else 'No explicit source/field declaration is available')


def imagery(status: str, checked_at: datetime, basis: str, reason: str, times=()) -> ImageryAvailability:
    return ImageryAvailability(status=status, checked_at=checked_at, basis=basis, reason=reason, times=list(times))

File: experiments/st-johns-weather-map/web/src/workbench/layerIdentity.ts
import type { LayerItem } from '../types'
import { resolveFieldKey } from '../fieldFamily'

export function layerMapping(layer: LayerItem) {
  const unknown = { status: 'unknown' as const, fields: [] as NonNullable<LayerItem['field_mappings']>, reason: layer.mapping_reason ?? 'Source identity not supplied; no join is inferred from product or title' }
  if (layer.mapping_status === undefined && layer.field_mappings === undefined) return unknown
  if (!['known', 'partial', 'unknown'].includes(layer.mapping_status ?? '') || !Array.isArray(layer.field_mappings) || layer.field_mappings.length > 128) return { ...unknown, reason: 'Source mapping is unreadable' }
  const fields = layer.field_mappings
  if (!fields.every((row) => row && typeof row.source_id === 'string' && row.source_id.length > 0 && (row.field_key === null || (typeof row.field_key === 'string' && resolveFieldKey(row.field_key) === row.field_key)) && (row.declared_field === null || typeof row.declared_field === 'string'))) return { ...unknown, reason: 'Source mapping is unreadable' }
  const expected = fields.length ? fields.every((row) => row.field_key !== null) ? 'known' : 'partial' : 'unknown'
  if (expected !== layer.mapping_status) return { ...unknown, reason: 'Source mapping status contradicts its records' }
  return { status: layer.mapping_status, fields, reason: layer.mapping_reason ?? 'Mapping reason not supplied' }
}

export function layerImagery(layer: LayerItem) {
  const unknown = { status: 'unknown' as const, checked_at: null, basis: 'not_supplied', times: [] as string[], reason: 'Imagery availability is unknown; listed sample times are not an image inventory' }
  const value = layer.imagery_availability
  if (!value) return unknown
  if (!['known', 'unknown', 'unavailable'].includes(value.status) || typeof value.reason !== 'string' || typeof value.basis !== 'string' || !Array.isArray(value.times)
    || !value.times.every((time) => typeof time === 'string' && Number.isFinite(Date.parse(time)))
    || (value.checked_at !== null && (typeof value.checked_at !== 'string' || !Number.isFinite(Date.parse(value.checked_at))))
    || (value.status === 'known' && value.checked_at === null) || (value.status !== 'known' && value.times.length > 0)) return { ...unknown, reason: 'Imagery availability is unreadable' }
  return value
}
